yajsapi/storage/ws-browser.ts

```diff
@@ -0,0 +1,234 @@
+import { StorageProvider, StorageProviderDataCallback } from "./provider";
+import { v4 } from "uuid";
+import { encode, toObject } from "flatbuffers/js/flexbuffers.js";
+import { getIdentity } from "../network/identity.js";
+import * as jsSha3 from "js-sha3";
+import { Logger, nullLogger } from "../utils/index.js";
+
+
+export interface WebSocketStorageProviderOptions {
+  yagnaOptions: {
+    apiKey: string;
+    basePath: string;
+  };
+  logger?: Logger
+}
+
+interface GsbRequest<T> {
+  id: string;
+  component: string;
+  payload: T;
+}
+
+interface GetMetadataRequest extends GsbRequest<void> {
+  component: "GetMetadata";
+}
+
+interface GetChunkRequest extends GsbRequest<{ offset: number, size: number }> {
+  component: "GetChunk";
+}
+
+type UploadChunkChunk = { offset: number, content: Uint8Array }
+type UploadChunkPayload = {
+  chunk: UploadChunkChunk;
+};
+
+interface UploadChunkRequest extends GsbRequest<UploadChunkPayload> {
+  component: "UploadChunk";
+}
+
+interface UploadFinishedRequest extends GsbRequest<{ hash: string }> {
+  component: "UploadFinished"
+}
+
+type GsbRequestPublishUnion = GetMetadataRequest | GetChunkRequest;
+type GsbRequestReceiveUnion = UploadFinishedRequest | UploadChunkRequest;
+
+type ServiceInfo = {
+  url: URL;
+  serviceId: string;
+}
+
+type GftpFileInfo = {
+  id: string;
+  url: string;
+}
+
+/**
+ * Storage provider that uses GFTP over WebSockets.
+ */
+export class WebSocketBrowserStorageProvider implements StorageProvider {
+  /**
+   * Map of open services (IDs) indexed by GFTP url.
+   */
+  private services = new Map<string, string>();
+  private logger: Logger;
+
+  constructor(private readonly options: WebSocketStorageProviderOptions) {
+    this.logger = options.logger ?? nullLogger();
+  }
+
+  close(): Promise<void> {
+    return this.release(Array.from(this.services.keys()));
+  }
+
+  init(): Promise<void> {
+    return Promise.resolve(undefined);
+  }
+
+  async publishData(data: Uint8Array): Promise<string> {
+    const fileInfo = await this.createFileInfo();
+
+    const ws = await this.createSocket(fileInfo, ["GetMetadata", "GetChunk"]);
+    ws.addEventListener("message", (event) => {
+      const req = toObject(event.data) as GsbRequestPublishUnion;
+      if (req.component === "GetMetadata") {
+        this.respond(ws, req.id, { fileSize: data.byteLength });
+      } else if (req.component === "GetChunk") {
+        this.respond(ws, req.id, {
+          content: data.slice(req.payload.offset, req.payload.offset + req.payload.size),
+          offset: req.payload.offset,
+        });
+      } else {
+        this.logger.warn(`[WebSocketBrowserStorageProvider] Unsupported message in publishData(): ${(req as GsbRequest<void>).component}`);
+      }
+    });
+
+    return fileInfo.url;
+  }
+
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```

Comment:
I don't think we should disable tslint in this way, let's just remove this parameter. TypeScript will allow it